underfs/hdfs/src/test/java/alluxio/underfs/hdfs/HdfsUnderFileSystemTest.java

```diff
@@ -63,4 +87,85 @@ public void prepareConfiguration() throws Exception {
     Assert.assertEquals(ufsConf.get(PropertyKey.UNDERFS_HDFS_IMPL), conf.get("fs.hdfs.impl"));
     Assert.assertTrue(conf.getBoolean("fs.hdfs.impl.disable.cache", false));
   }
+
+  class PreadSeekableStream extends FilterInputStream implements Seekable, PositionedReadable {
+
+    /**
+     * Creates a <code>FilterInputStream</code>
+     * by assigning the  argument <code>in</code>
+     * to the field <code>this.in</code> so as
+     * to remember it for later use.
+     *
+     * @param in the underlying input stream, or <code>null</code> if
+     *           this instance is to be created without an underlying stream.
+     */
+    protected PreadSeekableStream(InputStream in) {
+      super(in);
+    }
+
+    @Override
+    public int read(long l, byte[] bytes, int i, int i1) throws IOException {
+      return ((FSDataInputStream) in).read(l, bytes, i, i1);
+    }
+
+    @Override
+    public void readFully(long l, byte[] bytes, int i, int i1) throws IOException {
+      ((FSDataInputStream) in).readFully(l, bytes, i, i1);
+    }
+
+    @Override
+    public void readFully(long l, byte[] bytes) throws IOException {
+      ((FSDataInputStream) in).readFully(l, bytes);
+    }
+
+    @Override
+    public void seek(long l) throws IOException {
+      ((FSDataInputStream) in).seek(l);
+    }
+
+    @Override
+    public long getPos() throws IOException {
+      return ((FSDataInputStream) in).getPos();
+    }
+
+    @Override
+    public boolean seekToNewSource(long l) throws IOException {
+      return ((FSDataInputStream) in).seekToNewSource(l);
+    }
+  }
+  /**
+   * Tests the dynamic switching between pread and read calls to underlying stream.
+   */
+  @Test
+  public void verifyPread() throws Exception {
+    File file = mTemporaryFolder.newFile("test.txt");
+    FileUtils.writeByteArrayToFile(file, CommonUtils.randomBytes(4096));
+    FilterInputStream in = (FilterInputStream) mHdfsUnderFileSystem.open(file.getAbsolutePath(),
+        OpenOptions.defaults().setPositionShort(true));
+    FSDataInputStream dataInput = Whitebox.getInternalState(in, "in");
+    PreadSeekableStream stream = new PreadSeekableStream(dataInput);
+    PreadSeekableStream spyStream = spy(stream);
+    Whitebox.setInternalState(in, "in", spyStream);
+    in.read();
+    in.read();
+    in.read();
+    in.read();
+    in.skip(2);
+    in.skip(2);
+    in.read();
+    in.skip(2);
+    in.read();
+    verify(spyStream, never()).read(anyInt(), any(byte[].class), anyInt(), anyInt());
+    verify(spyStream, times(6)).read(any(byte[].class), anyInt(), anyInt());
+    in.skip(1000);
```

Comment:
nit: comment that `skip > MOVEMENT_LIMIT` causes pread. maybe even mark `MOVEMENT_LIMIT` `@VisibleForTesting` and re-use `2*MOVEMENT_LIMIT` to avoid changing the test if we tune the parameter.

Reply:
done